mips/exc.S

```diff
@@ -85,30 +96,39 @@ user_exc_enter:
         nop
 
         # Interrupts may be enabled here and that's ok.
-
-        # Disable interrupts till ERET.
-        di
+        LOAD_PCPU($t0)
+        lw      $t0, PCPU_CURTHREAD($t0)
+        lw      $t1, TD_FLAGS($t0)
+        andi    $t1, TDF_NEEDSWITCH
+        beqz    $t1, user_exc_leave
+        move    $a0, $sp
+        jal     exc_before_leave
+        nop
 
 user_exc_leave:
+        # Disable interrupts and fetch C0_STATUS.
+        di      $t0
+
         # Fetch the context from thread control block.
         LOAD_PCPU($k0)
         lw      $k0, PCPU_CURTHREAD($k0)
         addu    $k0, TD_UCTX
 
-        # Update status register held in user context (interrupt mask).
-        LOAD_REG($t0, SR, $k0)
-        li      $t1, ~SR_IMASK
-        and     $t0, $t1
-        mfc0    $t1, C0_STATUS
-        andi    $t1, SR_IMASK
+        # Turn on FPU, enter exception level, enter user mode
+        # and enable interrupts (last two applies only after ERET).
+        # TODO: What about interrupt mask?
```

Comment:
I can't find any, at least for now. If one day we'll decide to implement ultra-priority threads which should not be preempted by some insignificant interrupts in order to improve their performance, then maybe it will be worth to experiment with interrupt masks. But for now we'll be fine handling everything immediately.